README.md
Midwife
==============

Small program that can be used as the interpreter for shebang lines, that then executes the real interpreter based on its command line argument

It is small, fast, and written in portable C++.


ChangeLog
Note that `drbo', `Demetri Obenour' and `Demetrios Obenour' are all
the same person (the last is my legal name, the first my Git username,
and the middle my preferred name).

2014-12-19  Demetrios Obenour  <demetriobenour@gmail.com>

	* . Basically changed every file in the project. Nothing is the same! See Git log for details.

commit 39e19ce
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    FINALLY finalized the m4 code that writes the m4 in package.m4! -- for real this time

autogen.sh
configure.ac
rebuild.sh

commit 2948465
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    FINALLY finalized the m4 code that writes the m4 in package.m4!

configure.ac

commit dc811fa
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Nearly done with m4 work

configure.ac

commit 2e504e5
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Added support for 'make installcheck'

Makefile.am

commit 6de7175
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Using printf instead of AT_DATA to create script file; tests now work

tests/testsuite.at

commit 7809f7a
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Deleted unneeded whitespace and uneeded check for stdbool.h

configure.ac

commit e0a2c7b
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Added @configure_input@ to Makefile.am

Makefile.am

commit 1b86983
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Changed testsuite.at to include package.m4 and clarified copyright notice

tests/testsuite.at

commit d361f81
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    .gitignore'd package.m4 (autogenerated by configure.ac)

.gitignore

commit d748f5d
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Added package.m4 to MAINTAINERCLEANFILES

Makefile.am

commit 28a8f8c
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Finished m4 code that writes package.m4

configure.ac

commit 483e8a9
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-18

    Cleaned up m4 code in configure.ac

configure.ac

commit 5dd0d28
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Cleaned up the m4 code that generates package.m4

configure.ac

commit e6d4e2a
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    make distclean now cleans files generated by the testsuite

Makefile.am

commit 400c44b
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Autoconf now creates & updates package.m4
    
    Through m4_syscmd, running `autoconf' now automatically writes to package.m4

configure.ac

commit 27b0c8f
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Cleaning up remenants of old test system

Makefile.am
configure.ac

commit 7e06600
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Renamed LICENSE to COPYING

COPYING
LICENSE

commit af484a8
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Switched to Autotest test system

Makefile.am

commit 63206a3
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Deleted unneeded file

tests/test.in

commit 81e65f2
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Added autotest-generated files to .gitignore

.gitignore

commit 915a452
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Added script for rebuilding the Autoconf-generated files

rebuild.sh

commit 3150286
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Working test suite using GNU Autotest

tests/testsuite.at

commit 96a168b
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Fixed generation of @printf@ by configure

configure.ac

commit 79a0c04
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    Fixed Makefile.am

Makefile.am

commit 35558d5
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    This version of midwife.cpp builds and passes make check

src/midwife.cpp

commit 1107ddb
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    This version of the header has been tested

src/lexer.hpp

commit 40ded6f
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-17

    This version of the lexer has been tested to work
    
    This version of the lexer has been tested to work. It builds and passes make check.

src/lexer.cpp.in

commit e278ddc
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-16

    Working lexer! (hopefully)
    
    This version passes a basic test

src/lexer.cpp.in

commit 35ddd3c
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-16

    re2c requires that the code for no action fall off of the end
    
    In re2c, if the code for any action falls off of the end, the behavior is undefined. This cost me several hours debugging.

src/lexer.cpp.in

commit 1ba102e
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-15

    Creating my first test case

tests/test.in

commit 5c9d00b
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-15

    Gitignore'd generated files

.gitignore

commit dca5124
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-15

    Changed README.md to reflect that the program is now in C++

README.md

commit 7d7e230
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-15

    Fixed trivial change in #include
    
    No good reason to use #include "./lexer.hpp" instead of #include "lexer.hpp"

src/midwife.cpp

commit d22a720
Author: Demetri Obenour <demetriobenour@gmail.com>
Date:   2014-12-15

    Added src/lexer.hpp to source and distributed files
    

src/midwife.c
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#ifndef _POSIX_C_SOURCE
# define _POSIX_C_SOURCE 200809L
#endif

#ifndef __cplusplus
# include <stdbool.h>
# include <iso646.h>
#endif

#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>


#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
# define NORETURN _Noreturn
#elif defined(__cplusplus) && __cplusplus >= 201103L
# define NORETURN [[noreturn]]
#elif defined(__GNUC__) && __GNUC__ > 2 ||      \
  (__GNUC__ == 2 && __GNUC_MINOR__ >= 5)
# define NORETURN __attribute__((noreturn))
#elif !defined(NORETURN)
# define NORETURN
#endif

static NORETURN void die (const char *arg);
static void die(const char *arg) {
  fputs(arg, stderr);
  exit(127);
}

static char *program_name;
int parseHex(char a);
int parseHex(char a) {
  if ('0' <= a && a <= '9') {
    return a - '0';
  } else if ('a' <= a && a <= 'f') {
    int value = a - 'a';
    return value + 10;
  } else if ('A' <= a && a <= 'F') {
    int value = a - 'A';
    return value + 10;
  } else {
    die("Bad %- escape\n");
  }
}


int parseArgs(char *string, char ** buffer, char *dest);
int parseArgs(char *string, char ** buffer, char *dest) {
  char *cursor = dest;
  size_t numArg = 0;
  const char *ptr = string;
  buffer[numArg++] = dest;
  while(true) {
    char value = *ptr++;
    int result;
    switch (value) {
      case '&': *cursor++ = '\0';
        buffer[numArg++] = cursor;
        break;
      case '+': *cursor++ = ' ';
        break;
      case '%':
        result = 16 * parseHex(*ptr++);
        result += parseHex(*ptr++);
        *cursor++ = (result > 127) ? result - 256 : result;
        break;
      case '\0':
        *cursor = '\0';
        buffer[numArg] = (char *)0;
        return 0;
      default: *cursor++ = value;
        break;
    }
  }
}
void usage(void);
void usage(void) {
  fprintf(stderr, "usage: %s (percent-encoded argument) filename ...\n",
          program_name);
}
#if 0
void debugPrint(char ** buffer);
void debugPrint(char ** buffer) {
  char *string;
  while ((string = *buffer++)) {
    fputs(string, stderr);
  }
}
#endif

char * parseLines (char * numLines, char *fileName) {
  char *ptr = 0;
  char *end;
  {
    FILE *fileptr = fopen(fileName, "r");
    if ((FILE*)0 == fileptr) {
      die("Cannot open input file\n");
    }
    errno = 0;
    unsigned long int numlines = strtoul(numLines, &end, 10);
    if (0 != errno || '\0' != *end) {
      die("bad number of lines");
    }

    size_t length;
    off_t linesize;
    for (size_t i = 0; i < numlines; ++i) {
      linesize = getline(&ptr, &length, fileptr);
    }
    fclose(fileptr);
  }
  char *begin = strstr(ptr, "-*-");
  if (NULL == begin) {
    die("missing -*- start token");
  }
  else {
    begin += 3;
    char * end = strlen(begin) + begin - 1;
    for (char* i = end; i != begin; --i) {
      if (0 == strncmp(i, "-*-", 3)) { break; }
    }
    if (end == begin) {
      die("missing -*- end token");
    }
  }
  *end = '\0';
  return begin;
}

/*!re2c */

int main(int argc, char **argv) {
  char * dest[128];
  char **buffer;
  program_name = argv[0];
  size_t bufsize;
  if (argc < 3) {
    usage();
    return 127;
  }
  char *extras = 0;

  size_t numextraargs = 1;
  {
    char *ptr = argv[1];
    char value;
    if (0 == strncmp(ptr, "-n", 2)) {
      extras = parseLines(ptr + 2, argv[2]);
      numextraargs = 3;
    } else {
      while ((value = *ptr++)) {
        if ('&' == value) {
          ++numextraargs;
        }
      }
    }
  }


  /* Subtract one argument for our own name
   * and another for the command string */
  bufsize = (argc + numextraargs - 1);
  buffer = bufsize <= 128 ? dest : (char**)calloc(bufsize, sizeof(char *));
  if (0 == buffer) {
    die("Error allocating memory\n");
  }
  /* Parse arguments */
  if (extras) {
    buffer[0] = "/bin/sh";
    buffer[1] = "-c";
    buffer[2] = extras;
  } else {
    char *result = argv[1];
    int value = parseArgs(argv[1], buffer, result);
    if (value) {
      return value;
    }
  }
  fflush(stderr);
  memcpy(buffer + numextraargs, argv + 2, sizeof(char *) * (argc - 1));
  execv(buffer[0], buffer);
  die("Failed to exec!\n");
}


src/shebang-helper.c
#define _POSIX_C_SOURCE 200809L
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#ifdef __GNUC__
# define NORETURN(a) a __attribute__((noreturn))
#else
# if defined __STDC_VERSION__
#  if __STDC_VERSION__ >= 201112L
#   define NORETURN(a) _Noreturn a
#  endif
# endif
#endif
#ifndef NORETURN
# define NORETURN(a) a
#endif
#ifndef __cplusplus
# include <stdbool.h>
# include <iso646.h>
#endif
void NORETURN(die(const char *arg));
void die(const char *arg) {
  fputs(arg, stderr);
  exit(127);
}

static char *program_name;
int parseHex(char a);
int parseHex(char a) {
  if ('0' <= a && a <= '9') {
    return a - '0';
  } else if ('a' <= a && a <= 'f') {
    int value = a - 'a';
    return value + 10;
  } else if ('A' <= a && a <= 'F') {
    int value = a - 'A';
    return value + 10;
  } else {
    die("Bad %- escape\n");
  }
}


int parseArgs(char *string, char ** buffer, char *dest);
int parseArgs(char *string, char ** buffer, char *dest) {
  char *cursor = dest;
  size_t numArg = 0;
  const char *ptr = string;
  buffer[numArg++] = dest;
  while(true) {
    char value = *ptr++;
    int result;
    switch (value) {
      case '&': *cursor++ = '\0';
        buffer[numArg++] = cursor;
        break;
      case '+': *cursor++ = ' ';
        break;
      case '%':
        result = 16 * parseHex(*ptr++);
        result += parseHex(*ptr++);
        *cursor++ = (result > 127) ? result - 256 : result;
        break;
      case '\0':
        *cursor = '\0';
        buffer[numArg] = (char *)0;
        return 0;
      default: *cursor++ = value;
        break;
    }
  }
}
void usage(void);
void usage(void) {
  fprintf(stderr, "usage: %s (percent-encoded argument) filename ...\n",
          program_name);
}
void debugPrint(char ** buffer);
void debugPrint(char ** buffer) {
  char *string;
  while ((string = *buffer++)) {
    fputs(string, stderr);
  }
}

int main(int argc, char **argv) {
  char ** dest[128];
  char **buffer;
  program_name = argv[0];
  size_t bufsize;
  if (argc < 3) {
    usage();
    return 127;
  }
  else {
    size_t numextraargs = 1;
    char *result = 0;
    {
      char *ptr, value = '\0';
      for (ptr = argv[1], value = '\0'; (value = *ptr++);) {
        if ('&' == value) {
          ++numextraargs;
        }
      }
      result = argv[1];
    }


    /* Subtract one argument for our one name and another for the command string */
    bufsize = (argc + numextraargs - 1);
    buffer = bufsize <= 128 ? dest : malloc(bufsize * sizeof(char *));
    /* fprintf(stderr, "%zu\n", numextraargs); 
    // Parse arguments */
    { int value = parseArgs(argv[1], buffer, result); if (value) return value; }
    fflush(stderr);
    memcpy(buffer + numextraargs, argv + 2, sizeof(char *) * (argc - 1));
    execv(buffer[0], buffer);
  }
  die("Failed to exec!\n");
}


src/midwife.cpp
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif
#ifndef _POSIX_C_SOURCE
# define _POSIX_C_SOURCE 200809L
#endif
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <vector>
#include <string>

#include "lexer.hpp"


void die(const char *arg) {
  fputs(arg, stderr);
  exit(127);
}

static char *program_name;
int parseHex(char a);
int parseHex(char a) {
  if ('0' <= a && a <= '9') {
    return a - '0';
  } else if ('a' <= a && a <= 'f') {
    int value = a - 'a';
    return value + 10;
  } else if ('A' <= a && a <= 'F') {
    int value = a - 'A';
    return value + 10;
  } else {
    die("Bad %- escape\n");
  }
}


int parseArgs(char *string, char ** buffer, char *dest);
int parseArgs(char *string, char ** buffer, char *dest) {
  char *cursor = dest;
  size_t numArg = 0;
  const char *ptr = string;
  buffer[numArg++] = dest;
  while(true) {
    char value = *ptr++;
    int result;
    switch (value) {
      case '&': *cursor++ = '\0';
        buffer[numArg++] = cursor;
        break;
      case '+': *cursor++ = ' ';
        break;
      case '%':
        result = 16 * parseHex(*ptr++);
        result += parseHex(*ptr++);
        *cursor++ = (result > 127) ? result - 256 : result;
        break;
      case '\0':
        *cursor = '\0';
        buffer[numArg] = (char *)0;
        return 0;
      default: *cursor++ = value;
        break;
    }
  }
}
void debugPrint(const char ** buffer);

NORETURN void launch (stringVector stringVect);
void launch (stringVector stringVect) {
  size_t end = stringVect.size();
  const char ** newargs = new const char*[end + 1];
  for (size_t i = 0; i < end; ++i) {
    newargs[i] = stringVect[i].c_str();
  }
  newargs[end] = (char *)0;
  //  debugPrint(newargs);
  execv((const char *)newargs[0], (char * const *)newargs);
  die("Failed to exec!\n");
}
void usage(void);
void usage(void) {
  fprintf(stderr, "usage: %s (percent-encoded argument) filename ...\n",
          program_name);
}
void debugPrint(const char ** buffer) {
  const char *string;
  while ((string = *buffer++)) {
    fputs(string, stderr);
    fputc('\n', stderr);
  }
}

enum state { NOTHING, DASH, STAR_DASH };
NORETURN void parseLines (char * numLines, int argc, char **argv);
void parseLines (char * numLines, int argc, char **argv) {
  FILE *fileptr = fopen((char*)(argv[0]), "r");
  if ((FILE*)0 == fileptr) {
    die("Cannot open input file\n");
  }
  errno = 0;
  char *ptr = 0;
  unsigned long int numlines;
  {
    char *end;
    numlines = strtoul(numLines, &end, 10);
    if (0 != errno || '\0' != *end) {
      die("bad number of lines\n");
    }
  }

  size_t length;
  off_t linesize;
  for (size_t i = 0; i < numlines; ++i) {
    linesize = getline(&ptr, &length, fileptr);
  }
  char *begin = strstr(ptr, "-*-");
  if (NULL == begin) {
    die("missing -*- start token\n");
  }
  else {
    begin += 2;
    *begin = ' ';
    char * end = strlen(begin) + begin - 1;
    enum state status = NOTHING;
    for (char* i = end; i != begin; --i) {
      switch (status) {
        case NOTHING:
          if ('-' == *i) { status = DASH; }
          break;
        case DASH:
          if ('*' == *i) { status = STAR_DASH; }
          else if ('-' == *i) { status = DASH; }
          else { status = NOTHING; }
          break;
        case STAR_DASH:
          if ('-' == *i) {
            *i = '\0';
            launch(lexer(begin, i - begin, argc, argv));
          }
      }
    }
    die("No -*- end token found\n");
  }
}

int main(int argc, char **argv) {
  char * dest[128];
  char **buffer;
  program_name = argv[0];
  {
    char * index = program_name + strlen(program_name) - 1;
    if ('2' <= *index && '9' >= *index) {
      if (argc < 2) { 
        usage(); 
        return 127;
      }
      parseLines(index, argc - 1, argv + 1);
    }
  }
  size_t bufsize;
  if (argc < 3) {
    usage();
    return 127;
  }
  size_t numextraargs = 1;
  {
    char *ptr = argv[1];
    char value;
    if (0 == strncmp(ptr, "-n", 2)) {
      parseLines(ptr + 2, argc - 2, argv + 2);
    } else {
      while ((value = *ptr++)) {
        if ('&' == value) {
          ++numextraargs;
        }
      }
    }
  }


  /* Subtract one argument for our one name
   * and another for the command string */
  bufsize = (argc + numextraargs - 1);
  buffer = bufsize <= 128 ? dest : (char**)calloc(bufsize, sizeof(char *));
  if (0 == buffer) {
    die("Error allocating memory\n");
  }
  /* Parse arguments */
  {
    char *result = argv[1];
    int value = parseArgs(argv[1], buffer, result);
    if (value) {
      return value;
    }
  }
  fflush(stderr);
  memcpy(buffer + numextraargs, argv + 2, sizeof(char *) * (argc - 1));
  execv(buffer[0], buffer);
  die("Failed to exec!\n");
}
